6841 edit prefs

lache-melvin opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #6841

# 👩🏻‍💻 What does this PR do?

<!-- Explain the changes you made -->

Forgive me, she's a lil big 😅 

Adds a Manage UI for preferences:

![Screenshot 2025-03-28 at 10 21 45 AM](https://github.com/user-attachments/assets/7b428a3a-6fd4-47af-8f5b-05cd09dab6bb)

In development, we define a default. All values will be this default. User can set global value - all stores will then use this value, unless you define a store-override.

Values are autosaved.

<!-- why are the changes needed -->

<!-- Add a screenshot if there are UI changes  -->

## 💌 Any notes for the reviewer?

<!-- Do you have any specific questions for the reviewer? -->

These are just some example prefs to get the idea, I think we remove all except the `show_contact_tracing` one once happy?

Will add some inline comments too, but have added the store-pref select UI as a rough idea, I think we should hide/remove this for now, it probably needs to be paginated/searchable before it adds much value. Imagine in future we would also want to manage from the other direction (on stores page, select a store, can set all prefs) but I think this is the main one we need for now?

<!-- Is there a high risk/complicated change they should focus on? -->

<!-- any general areas of the codebase touched? any side effects caused? -->

<!-- Anything half cooked but going to be finished off in a different PR? -->

# 🧪 Testing

<!-- Explain the steps you'd take to test the changes of this PR manually -->

- [ ] On OMS Central
- [ ] There is now a Manage > Preferences page
- [ ] There is a list of available preferences
- [ ] Selecting one, you can set global pref
- [ ] It auto-saves, refreshing the page, the value persists
- [ ] Global pref now shows as the pref for all stores too
- [ ] Can set a pref per-store

# 📃 Documentation

- [x] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: <!-- _(e.g.)_ New `issued` column in `Requisitions` indicates stock quantity already in shipments -->
  1.
  2.


# 📃 Reviewer Checklist

The PR Reviewer(s) should fill out this section before approving the PR

**Breaking Changes**
- [ ] No Breaking Changes in the Graphql API
- [ ] Technically some Breaking Changes but not expected to impact any integrations

**Issue Review**
- [ ] All requirements in original issue have been covered
- [ ] A follow up issue(s) have been created to cover additional requirements

**Tests Pass**
- [ ] Postgres
- [ ] SQLite
- [ ] Frontend

lache-melvin commented on server/service/src/preference/preferences/complex_pref.rs:
```diff
@@ -1,15 +1,55 @@
-impl Preference<ComplexPref> for ComplexPref {
+pub struct ComplexOne;
+
+impl Preference for ComplexOne {
```
So in theory this is about as hard as it would get for us to define a new preference?!

lache-melvin commented on server/service/src/preference/preferences/mod.rs:
```diff
@@ -1,18 +1,35 @@
+    PreferenceRegistry {
+        show_contact_tracing: ShowContactTracing,
+        complex_one: ComplexOne,
+        months_of_stock: MonthsOfStock,
```
Then you just add your new pref to the registry here, and the `Preferences` struct below

lache-melvin commented on server/service/src/preference/preferences/mod.rs:
```diff
@@ -21,45 +38,33 @@ pub fn get_preferences(
     Ok(prefs)
 }
 
-pub trait Preference<T: Default + DeserializeOwned> {
```
moved to own file

lache-melvin commented on server/service/src/preference/preferences/show_contact_tracing.rs:
```diff
@@ -0,0 +1,11 @@
+pub struct ShowContactTracing;
+
+impl Preference for ShowContactTracing {
+    type Value = bool;
```
I think the majority of prefs will be booleans, so would look just like this :)

Chris-Petty commented:
Been in the DMs, let me know to come back?

lache-melvin commented:
@Chris-Petty come backkk

andreievg commented on client/packages/host/src/routers/ManageRouter.tsx:
```diff
@@ -6,33 +6,29 @@ const NameService = React.lazy(
-      '@openmsupply-client/system/src/IndicatorsDemographics/Service/Service'
-    )
+const ManageService = React.lazy(
+  () => import('@openmsupply-client/system/src/Manage/Service')
```
Can you please comment about this change, why it was done, etc.. (since PR is quite big already), happy to review this but a quick note 'This is renaming, this functionality is now 'common' etc..

lache-melvin replied:
Sorry did put comments as above but looks like they became outdated somehow, will add again 🙏

andreievg commented on client/packages/programs/src/JsonForms/common/components/Boolean.tsx:
```diff
@@ -21,7 +21,10 @@ const UIComponent = (props: ControlProps) => {
+      // On most json form elements, we want the label at the start as
+      // the input may span multiple lines. However, for a boolean field,
+      // the input is a single line, so we want the label centred.
+      sx={{ ...DefaultFormRowSx, alignItems: 'center !important' }}
```
What effect does this have on other toggles in json forms ?

lache-melvin replied:
Looks the same :)

![Screenshot 2025-03-31 at 4 16 54 PM](https://github.com/user-attachments/assets/4271f5c3-4a05-48a4-9342-5bad9b5fc75d)

lache-melvin commented on client/packages/host/src/routers/ManageRouter.tsx:
```diff
@@ -6,33 +6,29 @@ const NameService = React.lazy(
-  if (gotoIndicatorsDemographics) {
-    return <IndicatorsDemographicsService />;
+  if (goToManage) {
+    return <ManageService />;
```
Demographics was just one page under `Manage` section. Now that we are adding a second one, thought best to refactor "indicatorservice" to be manage service, and demographics just a page within it. Sorry for extra noise, it's all just moving files accordingly 🙏

andreievg commented on server/service/src/preference/upsert.rs:
```diff
@@ -0,0 +1,38 @@
+    pub value: String,
+}
+
+pub fn upsert_preference(
```
I wanted this to be type safe, and not at runtime but at compile time. 

The thought that I had was having upsert method for every preference, similar to the query method in Preference for now. And when we do by store preference or machine uuid we would probably need a different type of save, maybe with array of store -> pref objects.

andreievg commented on client/packages/system/src/Manage/Preferences/api/operations.graphql:
```diff
@@ -0,0 +1,40 @@
+query allPrefs {
```
so yeah the endpoints I was thinking about were like this:

```graphql

query MonthsOfStock {
    preference {
        monthsOfStock
    }
}
# Can query them individually, when needed
query ShowContractTracing {
    preference {
        showContractTracing
    }
}
# This preference has a global default but is also by store
query MonthsOfStockForStore($storeId:  String!) {
    preference {
        monthsOfStock(storeId: $storeId)
    }
}
# Saving
mutate SaveMonthsOfStock($globalDefault: String!) {
    preference {
        saveMonthsOfStock(default: $globalDefault)
    }
}
# Saving technically speaking we would need a way to mass configure these
mutate SaveMonthsOfStockForStores($input: MonthsOfStockForStoreUpsertInput!) {
    preference {
        saveMonthsOfStock(forStores: $input)
    }
}
``` 

Above input would be like this:

{ 
    value: string,
    storeId: string
}[]

lache-melvin replied:
At least for the query side... that is how it works? Sorry no frontend example yet, but that's what this node is:

https://github.com/msupply-foundation/open-msupply/blob/8974bc9619a1d709403e8dc4f9c6362f5ad24339/server/graphql/types/src/types/preferences.rs#L13-L25

 `allPrefs` is a special query only used to see full list of prefs in the edit view

I think mutation you have is also very similar to what is there, `input` object just has optional `store_id` to distinguish global and forStore. Are you saying you think its worth distinguishing the two inputs further?

Don't think it would be too hard to change mutation to accept an array of inputs if we wanted to do mass upserts in future

andreievg commented on client/packages/system/src/Manage/Preferences/api/usePreferencesGraphQL.ts:
```diff
@@ -0,0 +1,15 @@
+import { getSdk } from './operations.generated';
+
+export const usePreferencesGraphQL = () => {
+  const { client } = useGql();
```
Ias looking at this and I think we just need one generic hook for the whole app, used like:

const { sdk, storeId } = useGraphQL(getSdk); <- and this is used inside the hooks

andreievg commented on client/packages/system/src/Manage/Preferences/EditPage/EditPreferencesPage.tsx:
```diff
@@ -0,0 +1,35 @@
+  const selectedPref = data?.find(d => key === d.key);
+
+  return (
+    <LineEditBase currentKey={key ?? ''} prefs={data ?? []}>
```
I honestly think this should be a normal table, where we can see all preferences, and then we go to matrix view for individual preferences for stores etc..

andreievg reviewed (changes requested):
Hi, keen to have a chat about this in person.

I was thinking about JSON forms, and here is my position:

We should be using them for things that are not know at compile time, like the dynamic nature of some configurations, programs, custom reports etc.. 

Otherwise I think we should be using normal react.

I like the Preference type definition, and the load method for it (which can be overwritten), but the load method may require more parameters, like machine uuid etc.. I think the main benifit of the Preference type definition is to associate 'type', 'key', and default of the pref and to help with serialization and deserialization. 

Load should really be a bunch of helper methods, new one added when needs arrise.

lache-melvin commented:
Hi thanks for review and chat @andreievg  - per discussion:

- Removed extra example prefs: [025eb49](https://github.com/msupply-foundation/open-msupply/pull/7087/commits/025eb49c7a515ce5f6cb1ec5dd1e93602f37e43e) (I think worth coming back to this diff, with current pattern this is all the code required to add a new pref
- Next 4 commits are just some radical code deletion
  - Remove the per-store piece [631d8ce](https://github.com/msupply-foundation/open-msupply/pull/7087/commits/631d8ceeffb40ecae798cae9d7d73101395a1828)
  - As you thought, didn't need the `PreferenceDescription` trait abstraction, except for the `type Value` on each implementation... Seeing as only have `bool` type for now, I just hard coded value to remove the extra trait, as wanted to remove the extra code. I think we will remove the need for this description concept with more type secure edit UI anyway :) [45533e4](https://github.com/msupply-foundation/open-msupply/pull/7087/commits/45533e42be11cebc1b636b019eec9e79ea5e1cdf)
  - Remove prefs by key query - only doing global prefs for this release, so just loading the pref through main prefs query and displaying. Not sure where we are going with store pref UI, we can drive backend based on frontend requirements at the time :)
- Added TODOs throughout for during cooldown, I am keen to work on this 🔥  Will create issue

lache-melvin commented on client/packages/system/src/Manage/Preferences/EditPage/EditPreference.tsx:
```diff
@@ -0,0 +1,58 @@
+            value={value}
+            config={selected}
+            // TODO: backend would generate IDs, just send the value
+            onChange={value => update({ value, id: `${selected.key}_global` })}
```
Please note this change, if you have existing global prefs in DB, would need to clear

andreievg commented:
Thanks @lache-melvin, good to have our chat too
